### PR TITLE
Propagate `onHeaders` for gRPC client.

### LINKED ISSUE #4583: `ClientCall.Listener#onHeaders` is not invoked for gRPC client
Currently, the gRPC client does not invoke `ClientCall.Listener#onHeaders`.

This behavior is different from upstream, which eventually invokes this callback.

https://github.com/grpc/grpc-java/blob/f08300e0e314c6f0c0731c46f3e9985aefa35513/netty/src/main/java/io/grpc/netty/NettyClientStream.java#L336-L345

We should also note that this callback is called only for headers (not for trailers)

ref: https://line-armeria.slack.com/archives/C1NGPBUH2/p1671523590046549

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

`ClientCall.Listener#onHeaders(Metadata)` is not invoked for armeria unlike the upstream implementation.

1. The callback should be called after header validation is done
https://github.com/grpc/grpc-java/blob/f08300e0e314c6f0c0731c46f3e9985aefa35513/core/src/main/java/io/grpc/internal/Http2ClientStreamTransportState.java#L88-L108
2. The callback should be called after transport related headers are stripped
https://github.com/grpc/grpc-java/blob/f08300e0e314c6f0c0731c46f3e9985aefa35513/core/src/main/java/io/grpc/internal/Http2ClientStreamTransportState.java#L110
3. The callback shouldn't be called for trailer-only responses
(https://github.com/grpc/grpc-java/blob/f08300e0e314c6f0c0731c46f3e9985aefa35513/netty/src/main/java/io/grpc/netty/NettyClientStream.java#L336-L345)

Note: I decided not to do content-type validation for the implementation although the [specification](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses) requires it. The reasoning behind this was that I didn't want to waste cpu cycles doing a check that isn't really used.
Having said this, let me know if anyone feels like we should add the check.
https://github.com/grpc/grpc-java/blob/f08300e0e314c6f0c0731c46f3e9985aefa35513/core/src/main/java/io/grpc/internal/Http2ClientStreamTransportState.java#L221-L225

Modifications:

- Added a `transportReportHeaders` callback to `TransportStatusListener` which is invoked when a headers is receieved.
- The `TransportStatusListener#transportReportHeaders` callback is invoked when `HttpStreamDeframer#processHeaders` is invoked.
- `ArmeriaClientCall#transportReportHeaders` calls `listener.onHeaders` when invoked.

Result:

- Closes #4583

### CODE REVIEW

File: grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientListenerTest.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                       (ctx, req) -> HttpResponse.of(HttpStatus.NO_CONTENT));
+        }
+    };
+
+    private final AtomicReference<Metadata> trailersRef = new AtomicReference<>();
+    private final AtomicReference<Metadata> headersRef = new AtomicReference<>();
+    private final AtomicReference<Status> statusRef = new AtomicReference<>();
+    private final AtomicReference<SimpleResponse> responseRef = new AtomicReference<>();
+
+    private final Listener<SimpleResponse> listener = new Listener<SimpleResponse>() {
```
Comment (ikhoon) [inline grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientListenerTest.java:103]: nit: What do you think of converting this anonymous class into a static inner class and adding the atomic references to the member fields?

```java
class CapturingListener implements Listener<SimpleResponse> {
    @Nullable
    private Metadata headers;
    ...
}
```

**Review by ikhoon — APPROVED** [review]
Thanks, @jrhee17! 🙇‍♂️

File: grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/TransportStatusListener.java [hunk 2]
```diff
@@ -24,6 +24,9 @@
  * logic through this listener, and for clients the final response {@link Status} is also returned.
  */
 public interface TransportStatusListener {
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/TransportStatusListener.java:26]: This is not relevant to the change of this PR, but could you add `@FuntionalInterface` to the interface?
Could you also update the description?

**Review by minwoox — APPROVED** [review]
Just one nit. Thanks! 😄

**Pull request merged**